src/model/transaction/insertFragmentIntoContentState.js

```diff
@@ -13,121 +13,280 @@
+        [targetNextKey, 'prevSibling'],
+        lastRootFragmentBlockKey,
+      );
     }
 
-    var text = block.getText();
-    var chars = block.getCharacterList();
+    // update fragment parent links
+    fragmentRootBlocks.forEach(blockKey =>
+      blockMapState.setIn([blockKey, 'parent'], targetParentKey),
+    );
 
-    // Modify head portion of block.
-    var blockSize = text.length;
-    var headText = text.slice(0, targetOffset);
-    var headCharacters = chars.slice(0, targetOffset);
-    var appendToHead = fragment.first();
+    // update targetBlock parent child links
+    if (targetParentKey) {
+      const targetParent = blockMap.get(targetParentKey);
+      const originalTargetParentChildKeys = targetParent.getChildKeys();
 
-    var modifiedHead = block.merge({
-      text: headText + appendToHead.getText(),
-      characterList: headCharacters.concat(appendToHead.getCharacterList()),
-      type: headText ? block.getType() : appendToHead.getType(),
-      data: appendToHead.getData(),
-    });
+      const targetBlockIndex = originalTargetParentChildKeys.indexOf(targetKey);
+      const insertionIndex = targetBlockIndex + 1;
 
-    newBlockArr.push(modifiedHead);
+      const newChildrenArray = originalTargetParentChildKeys.toArray();
+
+      // insert fragment children
+      newChildrenArray.splice(insertionIndex, 0, ...fragmentRootBlocks);
+
+      blockMapState.setIn(
+        [targetParentKey, 'children'],
+        new List(newChildrenArray),
+      );
+    }
+  });
+};
+
+const insertFragment = (
+  contentState: ContentState,
+  selectionState: SelectionState,
+  blockMap: BlockMap,
+  fragment: BlockMap,
+  targetKey: string,
+  targetOffset: number,
+): ContentState => {
+  const isTreeBasedBlockMap = blockMap.first() instanceof ContentBlockNode;
+  const newBlockArr = [];
+  const fragmentSize = fragment.size;
+  const target = blockMap.get(targetKey);
+  const head = fragment.first();
+  const tail = fragment.last();
+  const finalOffset = tail.getLength();
+  const finalKey = tail.getKey();
+  const shouldNotUpdateFromFragmentBlock =
+    isTreeBasedBlockMap &&
+    (!target.getChildKeys().isEmpty() || !head.getChildKeys().isEmpty());
+
+  //
+  // TODO - improve the bellow comment
+  // the first fragment block's content is used to update the target block
+  // we should not skip the first fragment block since we cannot merge the
+  // target block with the first fragment block contents
+  const startFragmentOffset = shouldNotUpdateFromFragmentBlock ? 0 : 1;
+
+  // check here why the fragment block insertion seems messed up
+  // on the snapshots the 'root' block is not appearing correct order
+  blockMap.forEach((block, blockKey) => {
+    if (blockKey !== targetKey) {
+      newBlockArr.push(block);
+      return;
+    }
+
+    // update head
+    newBlockArr.push(
+      updateHead(
+        block,
+        targetOffset,
+        fragment,
+        shouldNotUpdateFromFragmentBlock,
+      ),
```

Comment:
We could refactor this to simplify that logic, the tricky part is that we can't just add the block since we also need to update its links, will try to think on how to address this without using a boolean flag, if it turns out to be too complex i may address this in follow up PRS

Reply:
will remove the the need to pass boolean flags around, I agree with you and they are not great solutions